### ISSUE TITLE
Support for existence threshold in read API calls

### ISSUE THREAD

**Issue by BradleyFactual** [Factual/factual-ios-sdk#5]
https://support.factual.com/factual/topics/how-to-set-threshold-for-existence-with-swift

This user is trying to add the _threshold_ parameter to a request. I think the _threshold_  parameter didn't exist at the time the driver was authored, so the solution they are trying to pursue here is to just add an arbitrary key value pair to the request from within Swift. Perhaps the simplest solution is to formally support the _threshold_ parameter in this driver rather than to provide the flexibility to ad-hoc add any key/value parameter into your requests. 

[Example of the existence parameter.](http://api.v3.factual.com/t/places-us?threshold=confident&KEY=1qlm9AcPwxbtb3jZnV7G98t6cM6OojtSGbvZ88ul)

Threshold takes values from an enumeration: **confident*, **default**, or **comprehensive**. It's an optional parameter. If you don't provide it, the system returns the default.

**Comment by BradleyFactual** [comment]
related ticket: https://github.com/Factual/issues/issues/34

**Comment by yoshimotob** [comment]
Copied from support.factual.com:

If the type of "queryObject" is FactualQuery, it doesn't take in key value pairs, but follows the interface here:
https://github.com/Factual/factual-ios-sdk/blob/master/FactualQuery.h

Since the interface also doesn't support setting a threshold, I think if you will need to construct a custom query.  Here is a code example of how to do that:
https://github.com/Factual/factual-ios-sdk/blob/master/FactualSDKTests/FactualSDKTests.m#L195-L197

Basically, rather than using a FactualQuery, use a NSMutableDictionary to pass in the params.  Let us know if this works, thanks!

**Comment by BradleyFactual** [comment]
Thanks Brandon!

**Comment by mbfactual** [comment]
Thanks @yoshimotob and @BradleyFactual ! It looks like the customer was able to try the alternative presented via the support forum but ran into a snag. Is there anything I can pass along that might be helpful to them?

**Comment by yoshimotob** [comment]
Ok, I responded on the forum with the information to move forward

**Comment by tahoecoop** [comment]
Hi guys, I was the customer who started that initial thread referenced at the top. So I've got the code working with using receivedRawResult:, however the results I get back are much different. Using the code Brandon suggested (posted right below), I only get 3 results back in receivedRawResult. However, using the old code, I get 50 results (as it should be). Any ideas why this code returns such fewer results? I would still love to get this threshold working, thanks again!!
-------------------------- 
<NEW CODE BRANDON SUGGESTED THAT ONLY RETURNS 3 RESULTS>

let filters = NSMutableDictionary()
let includesAny = NSMutableDictionary()
let categories: NSMutableArray = [347, 312, 338]

includesAny.setValue(categories, forKey: "$includes_any") filters.setValue(includesAny, forKey: "category_ids")
let coords = NSMutableArray()

coords.addObject(userLocation.coordinate.latitude)      coords.addObject(userLocation.coordinate.longitude)

let centerFilter = NSMutableDictionary()

centerFilter.setValue(coords, forKey: "$center")

centerFilter.setValue(5000, forKey: "$meters")

let geoFilter = NSMutableDictionary()

geoFilter.setValue(centerFilter, forKey: "$circle")

let params = NSMutableDictionary()

params.setValue("confident", forKey: "threshold")

params.setValue("50", forKey: "limit")

params.setValue(filters, forKey: "filters")

params.setValue(geoFilter, forKey: "geo")

params.setValue("$distance:asc", forKey: "sort")

_apiObject.get("t/places-us", params: params as [NSObject : AnyObject], withDelegate: self)

-------------------------

<MY OLD CODE THAT IS RETURNING 50 RESULTS, LESS THRESHOLD PARAMETER, WHICH I WOULD STILL LOVE TO IMPLEMNENT>

// setting up the factual query, using makeQuery() method from FactualQuery+MakeQuery.h

let queryObject:FactualQuery = FactualQuery.makeQuery()

// adding filter for category id 347 (restaurants), 312 (bars), and 338 (food & dining)

queryObject.addRowFilter(FactualRowFilter.fieldName("category_ids", includesAnyArray: ["347", "312", "338"]))

let coordinate:CLLocationCoordinate2D = CLLocationCoordinate2DMake(userLocation.coordinate.latitude, userLocation.coordinate.longitude)

queryObject.setGeoFilter(coordinate, radiusInMeters: 5000)

queryObject.limit = 50

let sortCriteria = FactualSortCriteria(fieldName: "$distance", sortOrder: FactualSortOrder_Ascending)

queryObject.primarySortCriteria = sortCriteria

_apiObject.queryTable("restaurants-us", optionalQueryParams: queryObject, withDelegate: self)

**Comment by BradleyFactual** [comment]
@tahoecoop, Do have the lat/lng being passed into the geofilter?

**Comment by tahoecoop** [comment]
I am using CLLocationCoordinate2D
  - latitude : 37.332331410000002
  - longitude : -122.0312186
Thanks!!!

**Comment by gateway** [comment]
@BradleyFactual Hey, we are still in a holding pattern on this and are not able to move forward with using the threshold feature which we hope helps out on the data return such as duplicates, places closed etc.. Also any ideas on when the sdk will be updates its been a few years.. 

And thank you for helping us.

**Comment by BradleyFactual** [comment]
I ran your params through the API directly, and I get 50 rows. Any chance you can put your code into debug mode and find out what the GET request it is generating looks like? My only guess is that the driver is somehow creating the wrong API query.

Debug mode: https://github.com/Factual/factual-ios-sdk#debug-mode

**Comment by tahoecoop** [comment]
Thanks for the response Bradley. Here is the console output from running both the queries. Using my old code that returns 50 results, I get:
2016-06-03 11:19:33.373 Likemoji[344:20924] Filter Queryt/restaurants-us?limit=25&sort=%24distance:asc&filters=%7B%22category_ids%22%3A%7B%22%24includes_any%22%3A%5B%22347%22%2C%22312%22%2C%22338%22%2C%22370%22%5D%7D%7D&geo=%7B%22%24circle%22%3A%7B%22%24center%22%3A%5B39.24949445299734%2C-119.9542478566244%5D%2C%22%24meters%22%3A5000%7D%7D
2016-06-03 11:19:33.376 Likemoji[344:20924] Url: http://api.v3.factual.com/t/restaurants-us?limit=25&sort=%24distance:asc&filters=%7B%22category_ids%22%3A%7B%22%24includes_any%22%3A%5B%22347%22%2C%22312%22%2C%22338%22%2C%22370%22%5D%7D%7D&geo=%7B%22%24circle%22%3A%7B%22%24center%22%3A%5B39.24949445299734%2C-119.9542478566244%5D%2C%22%24meters%22%3A5000%7D%7D
2016-06-03 11:19:33.376 Likemoji[344:20924] Request Method: GET
2016-06-03 11:19:33.377 Likemoji[344:20924] Payload: (null)
2016-06-03 11:19:33.377 Likemoji[344:20924] using OAuth Request for Places Request
2016-06-03 11:19:33.838 Likemoji[344:20924] NSURLConnection:1 URL:http://api.v3.factual.com/t/restaurants-us?limit=25&sort=%24distance:asc&filters=%7B%22category_ids%22%3A%7B%22%24includes_any%22%3A%5B%22347%22%2C%22312%22%2C%22338%22%2C%22370%22%5D%7D%7D&geo=%7B%22%24circle%22%3A%7B%22%24center%22%3A%5B39.24949445299734%2C-119.9542478566244%5D%2C%22%24meters%22%3A5000%7D%7D Finished Loading

Using the new code from Brandon that only returns 3 results, I only get in the console:
2016-06-03 11:27:01.460 Likemoji[368:24071] NSURLConnection:1 URL:http://api.v3.factual.com/t/places-us?geo=%7B%22%24circle%22%3A%7B%22%24center%22%3A%5B39.2494853586324%2C-119.9542367086932%5D%2C%22%24meters%22%3A5000%7D%7D&limit=50&filters=%7B%22category_ids%22%3A%7B%22%24includes_any%22%3A%5B347%2C312%2C338%5D%7D%7D&sort=%24distance%3Aasc&threshold=confident Finished Loading

Anyway, thanks again for your assistance!!

**Comment by BradleyFactual** [comment]
The query you are showing from 2016-06-03 11:27:01.460 returns 50 rows for me when I run it. How are you determining it to only have 3 results? Is it possible you are hitting an exceptional when iterating through the results?

**Comment by tahoecoop** [comment]
When it goes into  `public func requestComplete(request: FactualAPIRequest!, receivedRawResult result: [NSObject : AnyObject]!) { ...   }`, I stop it with a break point and I po result.count, which returns 3. However, I'm not getting the actual data for those places. It's really weird. I gave you my code above, how come you are getting different results than I am? Can you check my queries on your end? Thanks again!

**Comment by BradleyFactual** [comment]
Can you verify that your API key starts with nKPN90? When I look at the logs for that API key, I see queries that end in "threshold=confident".

I tried the following query out of your logs

http://api.v3.factual.com/t/places-us?geo={%22$circle%22:{%22$center%22:[39.24955341968615,-119.9543625210598],%22$meters%22:5000}}&limit=50&filters={%22category_ids%22:{%22$includes_any%22:[347,312,338]}}&sort=$distance:asc&threshold=confident

Running that query directly against our servers (just add KEY=your_api_key here to the query params, and you can run it in curl or your browser or whatever), I receive 50 rows of data.

**Comment by tahoecoop** [comment]
Yes that is our API key. The url you provided did work to get me the 50 results, but I that is only when we are using the web API. Through the iOS SDK however, I'm getting those weird results above. Do most of your customers use the web calls as opposed to the iOS SDK? Just trying to find the best/most reliable solution. I really appreciate it, thanks again!

**Comment by tahoecoop** [comment]
Interesting. I'm looking at these results in `receivedRawResult` before I parse the data at all, I'm just looking at the JSON. So it seems like the data may be getting lost somewhere in the SDK?

**Comment by BradleyFactual** [comment]
Can you output the raw JSON from receivedRawResult? Curious how many rows are in the JSON at that point.

**Comment by tahoecoop** [comment]
Well, I took out the debug code and I'm now getting the 50 places in the 3rd element of the data array. It seems like its good now, I'm not sure why I didn't see these results in there before, something I was doing probably? Anyway, I really appreciate the quick assistance from you!!! I'll let you know if I encounter any more issues, but seems like it's working. Have a great day!
